### ISSUE TITLE
com.gargoylesoftware.htmlunit.html.DomNode.asText() returns nested elements with linebreaks

### ISSUE THREAD

**Issue by Deadc00l** [HtmlUnit/htmlunit#315]
In the case of nested elements, the text of span element is output with linebreaks:
```html
<span>
    <select name="day" size="1">...</select>
    <select name="month" size="1">...</select>
    <select name="year" size="1">...</select>    
</span>
```

output: 
1
JAN
2021

Therefore, real line breaks, like < /br> cannot be distinguished in output text from blocks. This worked fine in the earlier versions like 2.18
It seems that the status is set incorrectly after a block, see:
com.gargoylesoftware.htmlunit.html.HtmlSerializer.HtmlSerializerTextBuilder.append(String, Mode)

**Comment by rbri** [comment]
Good point - can you please provide some test (in com.gargoylesoftware.htmlunit.html.HtmlSelectTest)

like this one

    /**
     * @throws Exception if an error occurs
     */
    @Test
    public void asTextInsideSpan() throws Exception {
        final String html = "<html><body><form>\n"
            + "<span id='test'>\n"
            + "  <select name='day' size='1'>"
                    + "<option value='1' selected>01</option>"
                    + "<option value='2'>02</option></select>\n"
            + "  <select name='month' size='1'>"
                    + "<option value='1' selected>Jan</option>"
                    + "<option value='2'>Feb</option></select>\n"
            + "  <select name='year' size='1'>"
                    + "<option value='1'>2022</option>"
                    + "<option value='2' selected>2021</option></select>\n"
            + "</span>"
            + "</form></body></html>";
        final HtmlPage page = loadPage(html);
        final DomElement elem = page.getElementById("test");
        assertEquals("01 Jan 2021", elem.asText());
    }

Just to make sure we are agree on the same expectations for this (simples selects, multiple selects, different size attributes etc.)

**Comment by rbri** [comment]
From my point of view the implementation for multiple selects is also not correct.

**Comment by rbri** [comment]
Ok, i think i have to make a hard cut here, will implement a new method asNormalizedText() and make the old one deprecated. Looks like there is a lot of stuff that is still there for backward compatibility.

same for #283

**Comment by Deadc00l** [comment]
Thank you. Here are some tests with expected values

/htmlunit/src/test/java/com/gargoylesoftware/htmlunit/html/HtmlTableTest.java:
```java
    /**
     * Tests getTableCell(int,int).
     * @exception Exception If the test fails
     */
    @Test
    public void getCellAt() throws Exception {
        final String htmlContent
            = "<html><head><title>foo</title></head><body>\n"
            + "<table id='table1' summary='Test table'>\n"
            + "<tr><td><a>cell1a</a><div><span>cell1b</span></div></td><td>cell2</td><td rowspan='2'>cell4</td></tr>\n"
            + "<tr><td colspan='2'>cell3</td></tr>\n"
            + "</table>\n"
            + "</body></html>";
        final HtmlPage page = loadPage(htmlContent);

        final HtmlTable table = page.getHtmlElementById("table1");

        final HtmlTableCell cell1 = table.getCellAt(0, 0);
        assertEquals("cell1 contents", "cell1acell1b", cell1.asText());

        final HtmlTableCell cell2 = table.getCellAt(0, 1);
        assertEquals("cell2 contents", "cell2", cell2.asText());

        final HtmlTableCell cell3 = table.getCellAt(1, 0);
        assertEquals("cell3 contents", "cell3", cell3.asText());
        assertSame("cells (1,0) and (1,1)", cell3, table.getCellAt(1, 1));

        final HtmlTableCell cell4 = table.getCellAt(0, 2);
        assertEquals("cell4 contents", "cell4", cell4.asText());
        assertSame("cells (0,2) and (1,2)", cell4, table.getCellAt(1, 2));
    }
    
    /**
     * Tests getTableCell(int,int).
     * @exception Exception If the test fails
     */
    @Test
    public void getCellAtWithBreaks() throws Exception {
        final String htmlContent
            = "<html><head><title>foo</title></head><body>\n"
            + "<table id='table1' summary='Test table'>\n"
            + "<tr><td><a>cell1a</a><div><br><span>cell1b</span></div></td><td>cell2</td><td rowspan='2'>cell4</td></tr>\n"
            + "<tr><td colspan='2'>cell3</td></tr>\n"
            + "</table>\n"
            + "</body></html>";
        final HtmlPage page = loadPage(htmlContent);

        final HtmlTable table = page.getHtmlElementById("table1");

        final HtmlTableCell cell1 = table.getCellAt(0, 0);
        assertEquals("cell1 contents", "cell1a\ncell1b", cell1.asText());

        final HtmlTableCell cell2 = table.getCellAt(0, 1);
        assertEquals("cell2 contents", "cell2", cell2.asText());

        final HtmlTableCell cell3 = table.getCellAt(1, 0);
        assertEquals("cell3 contents", "cell3", cell3.asText());
        assertSame("cells (1,0) and (1,1)", cell3, table.getCellAt(1, 1));

        final HtmlTableCell cell4 = table.getCellAt(0, 2);
        assertEquals("cell4 contents", "cell4", cell4.asText());
        assertSame("cells (0,2) and (1,2)", cell4, table.getCellAt(1, 2));
    }
```
/htmlunit/src/test/java/com/gargoylesoftware/htmlunit/html/HtmlSelectTest.java:
```java
    /**
     * @throws Exception if an error occurs
     */
    @Test
    public void asTextInsideSpan() throws Exception {
        final String html = "<html><body><form>\n"
            + "<span id='test'>\n"
            + "  <select name='day' size='1'>"
                    + "<option value='1' selected>01</option>"
                    + "<option value='2'>02</option></select>\n"
            + "  <select name='month' size='1'>"
                    + "<option value='1' selected>Jan</option>"
                    + "<option value='2'>Feb</option></select>\n"
            + "  <select name='year' size='1'>"
                    + "<option value='1'>2022</option>"
                    + "<option value='2' selected>2021</option></select>\n"
            + "</span>"
            + "</form></body></html>";
        final HtmlPage page = loadPage(html);
        final DomElement elem = page.getElementById("test");
        assertEquals("01Jan2021", elem.asText());
    }
```

**Comment by Deadc00l** [comment]
Hello @rbri,
I saw that there was a commit to my problem. Thank you very mutch for the quick implementation. When I ran my sample unit tests again with the asNormalizedText () method, I found that the problem for the tables had unfortunately not yet been resolved. Would it be possible to look at that again?

**Comment by rbri** [comment]
Hi @dennisduysak ,

this is work in progress, i have not done much on this. The plan is
* to introduce the new method and make the old deprecated (done)
* switch from system dependent line separator to \n (done)
* make unit tests for at least all the cases we have unit test for the old method
* make sure to remove all the special cases we have implemented in the old hierarchy - have test for all the cases and adapt the impl to something meaningful (at this step your case might be fixed)
* cleanup, add more tests and release

But please keep in mind i do all this in my spare time, having a real >40h per week job - means any help is welcome.

**Comment by Deadc00l** [comment]
Thank you for the update. I know it's hard to get involved in such a project alongside a full-time job.

**Comment by Deadc00l** [comment]
Hi @rbri ,
I saw that the changes for the select tags are already commited. Thanks a lot for this! 
However, the tables are still incorrectly formatted (see unit tests of HtmlTableTest.java in coment before). Is it possible to format them in the same way like the select tags?

**Comment by rbri** [comment]
@Deadc00l long long ago...

regarding the first table test - getCellAt()

    assertEquals("cell1 contents", "cell1acell1b", cell1.asNormalizedText());

There is a line break if the browser renders the table - I think your expectation is wrong (and the one in org.htmlunit.html.serializer.HtmlSerializerNormalizedText2Test.tableCell() is correct.

<img width="200" height="125" alt="Image" src="https://github.com/user-attachments/assets/c5666373-a29b-4a23-86c9-b91b453dadf0" />


And the same for the second test

    assertEquals("cell1 contents", "cell1a\ncell1b", cell1.asNormalizedText());

Browsers are rendering two line breaks.

<img width="198" height="155" alt="Image" src="https://github.com/user-attachments/assets/c1167766-abf9-48e2-ace0-b7558fede483" />

**Comment by rbri** [comment]
@Deadc00l will close this - please open a new one if you (still use HtmlUnit - hopefully) and you have something more to fix
